sdcm/send_email.py

```diff
@@ -122,16 +152,52 @@ def send_email(self, subject, content, html=True, files=()):
                              self.send_email, self.email_recipients)
 
     def send_report(self, results):
-        try:
-            email_data = self.build_data_for_render(results)
-            self.log.info('Send email with results to {}'.format(self.email_recipients))
-            html, attached_files = self.build_report(email_data)
-            self.send_email(subject=email_data['subject'], content=html, files=attached_files)
-        except Exception as details:  # pylint: disable=broad-except
-            self.log.error("Error during sending email: %s", details, exc_info=True)
+        report_data = self.build_data_for_report(results)
+        attachments_data = self.build_data_for_attachments(results)
+        for _ in range(10):
+            try:
+                report = self._generate_report(report_data)
+                attachments = self._generate_report_attachments(attachments_data)
+                self.send_email(subject=report_data['subject'], content=report, files=attachments)
+                self.log.info(f'Send email with results to {self.email_recipients}')
+                return
+            except (AttachementSizeExceeded, BodySizeExceeded) as exc:
+                report_data, attachments_data = self.cut_report_data(report_data, attachments_data, reason=exc)
+            except Exception as details:  # pylint: disable=broad-except
+                self.log.error("Error during sending email: %s", details, exc_info=True)
+
+    def _generate_report_attachments(self, attachments_data):
+        if attachments_data is None:
+            self.save_html_to_file(
+                {},
+                'attachment_excluded.html',
+                template_str='<html><body>Attachment was excluded due to the size limitation</body></html>')
+            return ('attachment_excluded.html',)
+        elif attachments_data:
+            return self.build_report_attachments(attachments_data)
+        else:
+            return tuple()
+
+    def _generate_report(self, report_data):
+        if report_data is None:
+            return self.render_to_html(
+                {},
+                '<html><body>Report was not sent due to the size limitation</body></html>')
+        elif report_data:
+            return self.build_report(report_data)
+        else:
+            return self.build_report(report_data)
+
+    def build_report(self, report_data, template_str=None):
+        return self.render_to_html(report_data, template_str=None)
 
-    def build_report(self, email_data):
-        return self.render_to_html(email_data), ()
+    def build_report_attachments(self, attachments_data, template_str=None):
```

Comment:
why this is needed?

Reply:
Idea is to sepparate generating of attachment from email body.
In this case we can have logic of gradualy cutting down size of email step by step: 
- reduce attachment size
- exclude attachments
- reduce body size
- finaly send simplified body that tells report wasn't sent due to the body limit

Otherwise it won't be possible or will be messy.